Objects/moduleobject.c

```diff
@@ -559,65 +559,88 @@ PyModule_GetState(PyObject* m)
 void
 _PyModule_Clear(PyObject *m)
 {
-    PyObject *d = ((PyModuleObject *)m)->md_dict;
-    if (d != NULL)
-        _PyModule_ClearDict(d);
+    PyModuleObject *module = (PyModuleObject *)m;
+    _PyModule_ClearDict(module->md_name, module->md_dict);
 }
 
 void
-_PyModule_ClearDict(PyObject *d)
+_PyModule_ClearDict(PyObject *module_name, PyObject *dict)
 {
+    int verbose = _Py_GetConfig()->verbose;
+    if (verbose) {
+        PySys_FormatStderr("# cleanup[3] wiping %S module\n", module_name);
+    }
+
+    // If the module has no dict: there is nothing to do.
+    if (dict == NULL) {
+        return;
+    }
+
     /* To make the execution order of destructors for global
        objects a bit more predictable, we first zap all objects
        whose name starts with a single underscore, before we clear
        the entire dictionary.  We zap them by replacing them with
        None, rather than deleting them from the dictionary, to
        avoid rehashing the dictionary (to some extent). */
 
-    Py_ssize_t pos;
-    PyObject *key, *value;
-
-    int verbose = _Py_GetConfig()->verbose;
+    for (int step=1; step <= 2; step++) {
```

Comment:
I think that it may be better now to clear all objects in the same order, without prioritizing underscored names.

Reply:
Oh. I didn't know, so I chose the conservative option: keep the existing "heuristic". I'm fine with making the cleanup simpler :-)